### ISSUE TITLE
Integration with mastodon social Mastodon / GNUSocial

### ISSUE THREAD

**Issue by fr1t2** [keybase/keybase-issues#2948]
See a definite need for this sweet service integrated into GNUsocial as well as mastodon.

**Comment by zQueal** [comment]
Added to #518

**Comment by Sigafoos** [comment]
I already did this in an unofficial way by [tooting a signed message](https://witches.town/@dconley/117846) but it would be great to be able to have it displayed on my Keybase profile (i.e. I had to search for that way back in my timeline)

**Comment by simon-brooke** [comment]
I've done the [tooting a signed message](https://social.yyy.scot/web/statuses/99772243106659276) hack by following the recipe [here](https://ryanmaynard.co/mastodon-keybase); but it would be really nice if this was integrated into the keybase.io web app.

**Comment by lx4r** [comment]
Is there any update on this? Mastodon seems to have gained some momentum in the last days ...

**Comment by KyleRConway** [comment]
I support this. I must say that one of the first things I did when joining keybase was to try to add my Mastodon account. Was disappointed I couldn't. And, as @eloquence notes above, this would help with verification of users across various mastodon instances.

**Comment by Laurelai** [comment]
We need to be able to integrate this

**Comment by jconnary** [comment]
Would love to see this happen.

**Comment by timnolte** [comment]
👍

**Comment by ianchanning** [comment]
There's a typo in the title: Should be **GNU** Social instead of GUN Social

**Comment by kenguest** [comment]
I also encourage this endeavour.

**Comment by lx4r** [comment]
@fr1t2 Would it suffice that there is an RSS feed of every Mastodon account (accessible by adding `.rss` to the profile URL, e.g. this is [mine](https://mstdn.io/@lx4r.rss))?

**Comment by Laurelai** [comment]
You can easily make public posts on mastodon if thats what you mean, you can view any public post without an account. There are also 4 slots for profile metadata that users can fill in that are displayed publicly on your profile.

**Comment by Laurelai** [comment]
Example: https://mastodon.starrevolution.org/@Laurelai/42447

**Comment by fr1t2** [comment]
Looking at this [post](https://github.com/keybase/keybase-issues/issues/3099) it seems like there is more to this than it seems.

**Comment by malgorithms** [comment]
We are strongly considering this as one of our next proof types, which we'll be coming back to shortly. The last one we added was facebook and we've mostly been focusing on improving keybase recently. But yes, we've been pretty flooded with mastodon requests. We'd consider doing it even before instagram, if it were easy.

Some integrations are easy for us, and some are difficult. There are 2 server components to a proof on keybase:

- the hunter
- the scraper/verifier

The hunter is executed after a user claims they've posted a proof on service X. For example, on twitter, when a user posts a proof, Keybase's servers' hunter _finds_ the tweet. (We can't expect the user to give us a link to it; they just tell us their username) . The hunter does this by looking through the user's recent tweets to find one that matches the proof. Depending on the service, this is a lot of work and prone to breakage. But at least it only has to happen once per proof.

Once the proof is found, it's verified by the scraper/verifier, and some quick lookup info (e.g., its URL) is stored in our database. The server then considers the proof valid.

Not that a client trusts the server!  When a client wants to identify the user, they don't need to do any of the hunting, they just get the proof link from the server and verify it. The client has code to make sure the proof is cryprographically valid and verifies it was posted by the correct user, often by some combination of the URL and either the JSON reply or HTML structure of the DOM. There's a lot to get right there.

It's absolutely critical the proof can be viewed by a user without an API key and _even if they're not logged into the mastodon instance_. This is something we've maintained for all our proof types.  So for mastodon, I'm guessing there are instances where people's profiles aren't public unless the viewer is logged in - for example as a test I just tried to go to a random profile on counter.social and it told me I had to log in.

So for keybase to do cross-instance mastodon proofs (which would be really cool) and feel very good that that our maintenance will be easy, and the proof will work for everyone, even those with higher privacy settings, we'd suggest the following very _small_ changes to mastodon:

- a keybase proof field on profile settings. this would be sort of like the 4 key/value settings that are on there now, except it'd just be a value, and by convention it would be publicly retrievable regardless of the instance/user settings. (people would only put these strings in if they were intended to be public anyway)
- some endpoint anyone could visit (preferably JSON) where the public could check this value for a given username, and it wouldn't require any API key
- this would show up on the mastodon profile for people allowed to look at the profile, with a link to the keybase profile (and of course keybase would link to it)
- ideally give us a landing URL so we could prefill these. For example, if we send a user to whatever.mastodon.site/username/fill_keybase?value=(whatever), and they're logged in, then they land on a page with it prefilled and they just have to hit submit. We do this with many of our partners.

I imagine that's very easy, but I don't know the politics of how mastodon distributes software to its instances. 

If all that worked, we could do multi-instance mastodon proofs in a BREEZE. And then all mastodon users would (1) have cryptographically-connected accounts, (2) be able to see them on profiles, and (3) have cryptographic sharing with each other inside the keybase app. And (4) keybase profiles would start promoting mastodon.

It would be pretty slick.

I just sent a DM to @Gargron on twitter but not sure if there's a better way to reach him, or if there's someone else I should talk to . Anyway, it would be a good match I think. And it would avoid all this workaround that people are talking about here.

**Comment by Gargron** [comment]
@malgorithms As long as you send an `Accept: application/json` header, the profile page will return JSON, no URL changing necessary. That's just an easy secondary way for me to link to the JSON in the browser. But yes all correct.

**Comment by malgorithms** [comment]
ah got it. ok, we'll be discussing this internally in the very near future. If someone on our team wants to connect to you (actually or I do) for some Q&A back and forth, what's best? Your zeonfederated email?

**Comment by schmittlauch** [comment]
regarding Webfinger: As this is not a core component of ActivityPub, should we really depend on webfinger here? AFAIK Mastodon currently is not 100% ActivityPub compliant by only federating with WebFinger aware AP servers.
WebFinger is quite for the current social architecture, but there might be [reasons to get rid of it](https://octodon.social/@cwebber/100579212913524773).

**Comment by aendrew** [comment]
I posted a thread about why this would be amazing here: https://mastodon.social/@aendrew/100590696008271245

tl;dr please please please make this happen. Keybase is the answer to like half of the complaints I hear people raise about switching to Mastodon from Twitter.

**Comment by schmittlauch** [comment]
@Gargron Don't worry, I don't want to stop this. I just got the impression from cwebber that stuff like HTTP signatures was left out for bureaucratic reasons, but tying accounts to an URI was done in an attempt (successful or not) to keep AP more general for other applications (e.g. calendars) or other more P2P like structures.
But if you think this will not hinder application of this to other AP servers or think it being unsignificant, go on then.

**Comment by xkr47** [comment]
Is the above procedure with webfinger something that could be ported to the rest of the ActivityPub-based fediverse as well? Or is there a need to elaborate something that can be standardized?
@Gargron, when dismissing AcitivityPub as a home for authentication matters, did it mainly refer to logging in users to the site and such or did it cover also identifying users etc on the site, like needed to exchange proof data? (Sorry for not completely grasping the context there..) What I'm thinking is; would it be a good idea to involve https://github.com/w3c/activitypub/ at some point to standardize this thing or should it be a separate standard? I think it would be good to have it standardized, perhaps even in a keybase-agnostic way.. No offense against keybase at all, it's great; just reflecting from the point of view of a potential standard..

**Comment by kaniini** [comment]
Please do not require Webfinger for this.

**Comment by RX14** [comment]
You don't need webfinger for this unless you want to support vanity domains anyway.

**Comment by shawneric** [comment]
It's been several months since any update on this topic has been given. @fr1t2 and @Gargron any updates from you guys regarding this project? Is vs 2.6.0rc1 the version to test this feature with?

**Comment by Gargron** [comment]
@shawneric I haven't heard anything outside of this GitHub issue. However Mastodon 2.6.0 will implement link ownership verification, independently of Keybase.

**Comment by malgorithms** [comment]
Sorry not to be posting here, but we've actually been working on this and are pretty far along, the goal being Mastodon (and other) support with minimal development. We'll have a development guide proposal ready shortly -- along with most of the code done on our side -- and if a site or mastodon instance follows the guide, they'll officially show up in Keybase's proofs list, and Keybase profiles (both in-app and on website) will link into that site and profile too. It's going to be really sweet and smoother than all our other proof types.

Also, the nice thing is it won't be like our traditional 3rd party proofs, where people can lie on the 3rd party. Verification before printing can go both ways. What I mean by this: if you see on Twitter, "Verifying myself, I am chriscoyne on keybase..." that might be a lie; Keybase wouldn't accept this connection, but of course Twitter _would_ print the claim.   But If a site follows the protocol we're proposing, then a user will (1) start on Keybase, (2) click the mastodon instance they want to prove, (3) land on the mastodon instance with just a button to click to make the connection, and (3) now the mastodon instance can simply say the Keybase username on profiles, and Keybase profiles can simply say the Mastodon instance and username. And both sides will link to each other, only if cryptographically verified by keybase. And Keybase apps will check this all without trusting the servers.

It'll be great for tying together mastodon users across multiple instances, if desired. Or proving mutual ownership between Twitter and Mastodon. 

Anyway, we're almost there with a proposal and very far along in the code too.

**Comment by nealmcb** [comment]
> @shawneric .... Mastodon 2.6.0 will implement link ownership verification, independently of Keybase.

Thanks.  Is this the feature / code you're talking about?

[tootsuite/mastodon: Redesign forms, verify link ownership with rel="me" by Gargron · Pull Request #8703 · ](https://github.com/tootsuite/mastodon/pull/8703)

How might it relate to the protocol @malgorithms just described?

**Comment by Gargron** [comment]
>How might it relate to the protocol @malgorithms just described?

That *is* the feature I was talking about. It's more like an alternative. [See also](https://wiki.zegnat.net/media/the-real-deal-about-rel-me.html)

**Comment by BenLubar** [comment]
Important: to find a Mastodon user, you can't just go via the domain name of their address. The `/.well-known/host-meta` file has the URL template for the webfinger API, so to get to my account, you'd need to do the following requests:

- https://lubar.me/.well-known/host-meta
- https://mastodon.lubar.me/.well-known/webfinger?resource=acct:ben@lubar.me
- https://mastodon.lubar.me/users/ben

**Comment by xkr47** [comment]
I think @malgorithms and @Gargron should set up a joint account at https://patreon.com/ and start collecting (monthly) money from all the hundreds of people that want this integration to happen! At least I would happily support it!

**Comment by malgorithms** [comment]
no Patreon needed from our perspective. We've been working on this! members of our team have actually been working on the Mastodon side of the integration and should have something to announce soon...

**Comment by malgorithms** [comment]
To clarify, Keybase isn't on Patreon.  But if you want to support Mastodon on Patreon, that's recommended!  https://www.patreon.com/mastodon

**Comment by JonahAragon** [comment]
Looks like Mastodon is just about ready to go with this feature. Have there been any updates from the Keybase team as to when this'll go live on their end?

**Comment by fr1t2** [comment]
Closing this issues as I have just verified myself in mastodon. Great job integrating these services! Open Source at its finest.

https://mastodon.social/@just_a_warning/101983826900173420

**Comment by rugk** [comment]
One issue still: The "proof" link here:
![image](https://user-images.githubusercontent.com/11966684/56726125-7afcab00-674e-11e9-96d3-760f0abea43b.png)

…does not actually link to the Mastodon toot that one can toot optionally, but just to the user profile. IMHO, this is not how it should be.

**Comment by rugk** [comment]
Okay, opened a new issue for that: https://github.com/keybase/keybase-issues/issues/3397

**Comment by Gargron** [comment]
> …does not actually link to the Mastodon toot that one can toot optionally, but just to the user profile. IMHO, this is not how it should be.

This is a misconception. The post is *optional* and is *not* the actual proof. Linking to the profile is the correct thing to do. The proof is saved directly in Mastodon, not inside a post!

**Comment by rugk** [comment]
…but this is not at all obvious to the user.

As I said repeatedly, the technical details are likely like that, but as a user the thing I am seeing is an optional toot and some profile on a click.

As an idea, maybe when we click the link from keybase, could Mastodon somehow highlight the tick?
So it is at least obvious to the user that this is the proof. Otherwise, i, as a user, just see the profile of Mastodon, okay, I am not sure where or what the proof right now is…

Please better refer/reply in https://github.com/keybase/keybase-issues/issues/3397.
